FI-2266: inferno new generator

Shaumik-Ashraf opened the pull request:
# Summary

Implemented working `inferno new` command with the following:

**Usage was updated [in comments below](https://github.com/inferno-framework/inferno-core/pull/408#issuecomment-1834502812)**

```
Usage:
  inferno new TEST_KIT_NAME [-i IG_URL]

Options:
  -i, [--implementation-guide=IG_URL]  # URL to an implementation guide or absolute path to a package.tgz
  -a, [--author=AUTHOR]                # Author names for gemspec file; you may use '-a' multiple times

Runtime options:
  -f, [--force]                    # Overwrite files that already exist
  -p, [--pretend], [--no-pretend]  # Run but do not make any changes
  -q, [--quiet], [--no-quiet]      # Suppress status output
  -s, [--skip], [--no-skip]        # Skip files that already exist

Generate a new Inferno test kit for FHIR software testing

Examples:

  `inferno new test_fhir_app`
    => generates an Inferno app

  `inferno new test_us_core -i https://build.fhir.org/ig/HL7/US-Core/`
    => generates Inferno app and loads US Core implementation guide

  `inferno new test_my_ig -i /absolute/path/to/ig/package.tgz -a Name`
    => generates Inferno app, loads a local implementation guide, and specifies Name as author

https://inferno-framework.github.io/index.html
```

Some key things to note in the implementation:
 - Made rails-like generator heavily leveraging `Thor::Actions`
 - /lib/inferno/apps/cli/new/templates dictates the contents of a new Inferno project
 - Added dry/inflector for good naming, but doesn't support acronyms like 'FHIR' well. No acronym support in this PR.
 - I tried to do rspec testing on it
 - There is a library called [aruba](https://github.com/cucumber/aruba) that can integrate with RSpec and test CLIs better, but I opted not to use it for now

# Testing Guidance

Both `rake spec` and `rubocop` should pass. The main testing file is `spec/inferno/apps/cli/new_spec.rb` which only focuses on testing naming and IG loading. Features from `Thor::Actions` like `--pretend` or `say` aren't tested via RSpec. 

For manual testing I did:
1. `mkdir tmp`
2. `cd tmp`
3. `bundle init`
4. add to Gemfile:
```
gem "inferno_core", path: "your/path/to/inferno_core/dev/with/fi-2266-inferno-new/branch"
```
5. `bundle install --binstubs`
6. `bin/inferno new --help` and other new commands

Jammjammjamm commented on lib/inferno/apps/cli/new.rb:
```diff
@@ -0,0 +1,74 @@
+      end
+
+      def authors
+        options['author'].empty? ? options['author'].append(default_author) : options['author']
```
```suggestion
        options['author'].presence || [default_author]
```

Jammjammjamm commented on lib/inferno/apps/cli/templates/%library_name%.gemspec.tt:
```diff
@@ -0,0 +1,26 @@
+  spec.description   = '<%= human_name %> Inferno test kit for FHIR'
+  # spec.homepage      = 'TODO'
+  spec.license       = 'Apache-2.0'
+  spec.add_runtime_dependency 'inferno_core', '~> 0.4.4'
```
This should be dynamically set to the current version.

Shaumik-Ashraf replied:
fixed

Jammjammjamm commented on lib/inferno/apps/cli/templates/README.md.tt:
```diff
@@ -0,0 +1,199 @@
+- [Ruby API documentation](https://inferno-framework.github.io/inferno-core/docs)
+- [JSON API documentation](https://inferno-framework.github.io/inferno-core/api-docs)
+
+## Instructions for Developing tests
```
These details have been moved out of the template readme, so this should be updated to reflect the current state of the readme.

Shaumik-Ashraf replied:
fixed, but I made a slight edit to the README mentioning `inferno new`

Jammjammjamm commented on lib/inferno/apps/cli/templates/spec/%lib_name%/patient_group_spec.rb.tt:
```diff
@@ -0,0 +1,105 @@
+RSpec.describe <%= module_name %>::PatientGroup do
```
`%lib_name%` in the path isn't being updated:
```
ᐅ git status
...
Untracked files:
  ...
	spec/%lib_name%/
```

Shaumik-Ashraf replied:
fixed

Jammjammjamm commented on lib/inferno/apps/cli/new.rb:
```diff
@@ -0,0 +1,74 @@
+        directory('.', root_name, { mode: :preserve, recursive: true, verbose: !options['quiet'] })
+        say_unless_quiet "Created #{root_name} Inferno test kit!", :green
+
+        return unless options['pretend']
```
Is this documented anywhere?

Jammjammjamm reviewed (commented):
I like the inclusion of the `put_ig_package_dot_tgz_here` file just to make things even more obvious.

I feel like we should probably include Gemfile.lock, although that seems like it could be tricky. Any ideas on a good way to do that?

Shaumik-Ashraf commented:
@Jammjammjamm ready for re-review.

For Gemfile.lock, I put in an Gemfile.lock.tt template that works in the current PR, but this solution feels flimsy. If inferno_core adds new dependencies `bundle install` will break on new templates, unless the team remembers to update the Gemfile.lock.tt template each time. There aren't any integration tests to catch this either.

Jammjammjamm commented:
> `rails new` just runs bundle install after the new generator. Should I implement this instead?

I think that makes sense. The user will have to run it anyway, so we may as well cut out that step.

Shaumik-Ashraf commented:
I replaced Gemfile.lock.tt template with running `bundle install`.

I also added a publish_template.sh script that puts up a new branch at inferno-template with the latest `inferno new inferno-template` run. Merging it shouldn't delete NOTICE.md from inferno-framework/inferno-template#22 (data rights legend PR) but does add `copyright 2024 TODO` back in the README.md. Is this okay?

Ready for review @Jammjammjamm, @arscan

Jammjammjamm reviewed (commented):
Could you just remove the publish script? This is so close to being done I don't think we should be adding anything new.

When I run `bin/inferno new inferno_template` I get:
```
inferno/lib/inferno/apps/cli/templates/%library_name%.gemspec.tt:11:in `template': uninitialized constant Inferno::VERSION (NameError)
```

Jammjammjamm commented:
After `bundle install`, this should run the migrations too. Otherwise I think this is good to go.

Shaumik-Ashraf commented:
Added `inferno migrate`

Jammjammjamm commented on lib/inferno/apps/cli/new.rb:
```diff
@@ -0,0 +1,103 @@
+        return if options['skip_bundle']
+
+        inside(root_name) do
+          run 'inferno migrate', verbose: !options['quiet'], capture: options['quiet']
```
```suggestion
          run 'bundle exec inferno migrate', verbose: !options['quiet'], capture: options['quiet']
```
